events/events.go

```diff
@@ -0,0 +1,379 @@
+	SetCreatedAt(*strfmt.DateTime)
+
+	UpdatedAt() *strfmt.DateTime
+	SetUpdatedAt()
+
+	Source() *string
+	SetSource(*string)
+
+	IPAddress() string
+	SetIPAddress(string)
+}
+
+// BaseBody contains data associated with all events.
+type BaseBody struct {
+	EventType       Type        `json:"type"`
+	StructActorID   manifold.ID `json:"actor_id"`
+	StructScopeID   manifold.ID `json:"scope_id"`
+	StructRefID     manifold.ID `json:"ref_id"`
+	StructCreatedAt time.Time   `json:"created_at"`
+	StructUpdatedAt time.Time   `json:"updated_at"`
+	StructSource    SourceType  `json:"source"`
+	StructIPAddress string      `json:"ip_address"`
+}
+
+// Validate returns an error if the BaseEventBody is not valid
+func (b *BaseBody) Validate(v interface{}) error {
+	// TODO: luiz - add validation
+	return nil
+}
+
+// Type returns the body's EventType
+func (b *BaseBody) Type() Type {
+	return b.EventType
+}
+
+// SetType sets the body's EventType
+func (b *BaseBody) SetType(s string) {
+	b.EventType = Type(s)
+}
+
+// ActorID returns the body's ActorID
+func (b *BaseBody) ActorID() manifold.ID {
+	return b.StructActorID
+}
+
+// SetActorID sets the body's ActorID
+func (b *BaseBody) SetActorID(id manifold.ID) {
+	b.StructActorID = id
+}
+
+// ScopeID returns the body's ScopeID
+func (b *BaseBody) ScopeID() manifold.ID {
+	return b.StructScopeID
+}
+
+// SetScopeID sets the body's ScopeID
+func (b *BaseBody) SetScopeID(id manifold.ID) {
+	b.StructScopeID = id
+}
+
+// RefID returns the body's RefID
+func (b *BaseBody) RefID() manifold.ID {
+	return b.StructRefID
+}
+
+// SetRefID sets the body's RefID
+func (b *BaseBody) SetRefID(id manifold.ID) {
+	b.StructRefID = id
+}
+
+// CreatedAt returns the body's CreatedAt
+func (b *BaseBody) CreatedAt() *strfmt.DateTime {
+	t := strfmt.DateTime(b.StructCreatedAt)
+	return &t
+}
+
+// SetCreatedAt sets the body's CreatedAt
+func (b *BaseBody) SetCreatedAt(t *strfmt.DateTime) {
+	if t == nil {
+		b.StructCreatedAt = time.Now().UTC()
+	} else {
+		b.StructCreatedAt = time.Time(*t)
+	}
+}
+
+// UpdatedAt returns the body's CreatedAt
+func (b *BaseBody) UpdatedAt() *strfmt.DateTime {
+	t := strfmt.DateTime(b.StructUpdatedAt)
+	return &t
+}
+
+// SetUpdatedAt sets the body's CreatedAt
+func (b *BaseBody) SetUpdatedAt() {
+	b.StructUpdatedAt = time.Now().UTC()
+}
+
+// Source returns the body's Source
+func (b *BaseBody) Source() *string {
+	s := string(b.StructSource)
+	return &s
+}
+
+// SetSource sets the body's Source
+func (b *BaseBody) SetSource(s *string) {
+	if s == nil {
+		b.StructSource = SourceSystem
+	} else {
+		b.StructSource = SourceType(*s)
+	}
+}
+
+// IPAddress returns the body's IPAddress
+func (b *BaseBody) IPAddress() string {
+	return b.StructIPAddress
+}
+
+// SetIPAddress sets the body's IPAddress
+func (b *BaseBody) SetIPAddress(ip string) {
+	b.StructIPAddress = ip
+}
+
+// OperationProvisioned represents a provision operation event.
+type OperationProvisioned struct {
+	BaseBody
```

Comment:
Should this be the other way around?
Since BaseBody contains more general information, that can will be more readily available, shouldn't it be at the top, then we drill down into more detailed stuff like OperationProvisionedData?

Reply:
FYI: https://golang.org/doc/effective_go.html#embedding we are embedding the basebody into the `OperationProvisioned` struct. Similar as we do with operations.